### PR TITLE
nhit promotion policy implementation

### PR DESCRIPTION
**Pull request by imjfckm**
Signed-off-by: Jan Musial <jan.musial@intel.com>

### CODE REVIEW

File: src/promotion/nhit/nhit.c [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+{
+	struct nhit_policy_context *ctx;
+	int result = 0;
+
+	ctx = env_vmalloc(sizeof(*ctx));
+	if (!ctx) {
+		result = -OCF_ERR_NO_MEM;
+		goto exit;
+	}
+	policy->ctx = ctx;
```
Comment (mmichal10) [inline src/promotion/nhit/nhit.c]: How about moving this assignment to line 45? Context is fully initialized at that point.

File: src/promotion/nhit/nhit.c [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+ocf_error_t nhit_set_param(ocf_promotion_policy_t policy, uint8_t param_id,
+		uint64_t param_value)
+{
+	struct nhit_policy_context *ctx = policy->ctx;
+	ocf_error_t result = 0;
+
+	switch (param_id) {
+	case nhit_insertion_threshold:
+		if (param_value < NHIT_MAX_THRESHOLD ||
```
Comment (mmichal10) [inline src/promotion/nhit/nhit.c]: Shouldn't condition looks like 'if ( param_value > NHIT_MIN_THRESHOLD && param_value < NHIT_MAX_THRESHOLD)'

File: src/promotion/nhit/nhit.c [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+				param_value < NHIT_MIN_THRESHOLD) {
+			env_atomic_set(&ctx->insertion_threshold, param_value);
+		} else {
+			ocf_cache_log(policy->owner, log_err, "Invalid nhit"
```
Comment (mmichal10) [inline src/promotion/nhit/nhit.c]: Missing space between 'nhit' and 'promotion' in next line.

File: src/promotion/nhit/nhit.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	return result;
+}
+
+void nhit_deinit(ocf_promotion_policy_t policy)
+{
+	struct nhit_policy_context *ctx = policy->ctx;
+
+	hash_deinit(ctx->hash_map);
+
+	env_vfree(ctx);
```
Comment (mmichal10) [inline src/promotion/nhit/nhit.c]: After you freed memory, set pointer value to NULL.

File: src/promotion/nhit/nhit.c (hunk 1 above)
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: True, fixed

File: src/promotion/nhit/nhit.c (hunk 2 above)
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: 😳

File: src/promotion/nhit/nhit.c (hunk 3 above)
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: Fixed

File: src/promotion/nhit/nhit.c (hunk 4 above)
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: Fixed

File: src/promotion/nhit/hash.c [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+	}
+
+	return needle;
+}
+
+static inline uint64_t get_rb_slot(nhit_hash_t ctx)
+{
+	env_spinlock_lock(&ctx->rb_pointer_lock);
+
+	ctx->rb_pointer = (ctx->rb_pointer + 1) % ctx->rb_entries;
```
Comment (mmichal10) [inline src/promotion/nhit/hash.c]: Please store this value in some temporary variable, don't return it directly.

File: src/promotion/nhit/hash.c [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+
+	new_ctx->ring_buffer = env_vzalloc(
+			new_ctx->rb_entries * sizeof(*new_ctx->ring_buffer));
+	if (!new_ctx->ring_buffer) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_locks;
+	}
+	for (i = 0; i < new_ctx->rb_entries; i++) {
+		new_ctx->ring_buffer[i].core_id = OCF_CORE_ID_INVALID;
+		env_atomic_set(&new_ctx->ring_buffer[i].counter, 0);
```
Comment (mmichal10) [inline src/promotion/nhit/hash.c]: Why you don't initialize ring_buffer[i].coll_prev and ring_buffer[i].coll_next with invalid values?

File: src/promotion/nhit/hash.c (hunk 5 above)
Comment (imjfckm) [reply src/promotion/nhit/hash.c]: Fixed

File: src/promotion/nhit/hash.c (hunk 6 above)
Comment (imjfckm) [reply src/promotion/nhit/hash.c]: I wanted to make sure the 'empty' state of slot is signified by a single variable just not to complicate the state

File: src/promotion/nhit/hash.c [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+	env_spinlock_init(&new_ctx->rb_pointer_lock);
+	new_ctx->rb_pointer = 0;
+
+	*ctx = new_ctx;
+	return 0;
+
+dealloc_locks:
+	env_vfree(new_ctx->hash_locks);
+	for (i = 0; i < new_ctx->hash_entries; i++)
+		ENV_BUG_ON(env_rwsem_destroy(&new_ctx->hash_locks[i]));
```
Comment (micrakow) [inline src/promotion/nhit/hash.c]: shouldn't it be done before freeing has_locks?

Comment (imjfckm) [reply src/promotion/nhit/hash.c]: Fixed, thanks

File: src/promotion/nhit/hash.c [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+			result = -OCF_ERR_UNKNOWN;
+			goto dealloc_locks;
+		}
+	}
+
+	new_ctx->ring_buffer = env_vzalloc(
+			new_ctx->rb_entries * sizeof(*new_ctx->ring_buffer));
+	if (!new_ctx->ring_buffer) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_locks;
```
Comment (arutk) [inline src/promotion/nhit/hash.c]: You should probably add deinit_rwsem label here ...

File: src/promotion/nhit/hash.c [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+ *	    |        +           |        |     |
+ *	    v        v           v        |     v
+ *	+--++-+--+---+-+--+--+--++-+--+--++-+--++-+
+ *	|  |  |  |  |X |  |  |  |  |  |  |  |  |  | ring_buffer
+ *	+--++-+--+---+-+--+--+--++-+--+--++-+--+--+
+ *	    |        ^           |        ^
+ *	    |________|           |________|
+ *
+ * Since rb_pointer is pointing to occupied rb slot we need to write-lock hash
+ * bucket I associated with this slot and remove it from collision list.
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply src/promotion/nhit/hash.c]: Got some more patches lined up after this PR, will fix it there!

File: src/promotion/nhit/hash.c (hunk 8 above)
Comment (imjfckm) [reply src/promotion/nhit/hash.c]: Got some more patches lined up after this PR, will fix it there!

Comment (imjfckm) [reply src/promotion/nhit/hash.c]: Got some more patches lined up after this PR, will fix it there!

**Pull request merged**